### PR TITLE
[FLINK-7515][network] allow actual 0-length content in NettyMessage#allocateBuffer()

### PR DESCRIPTION
**Pull request by NicoK**
## What is the purpose of the change

Previously, length "0" meant "unknown content length" but there are cases where the actual length is "0" and we do not need a larger buffer than the header size.

## Brief change log

- use -1 for tagging the special case of an unknown content length
- change buffers with 0-size content to start with the header size only

## Verifying this change

This change is already covered by existing tests, such as `NettyMessageSerializationTest` and many more network-related tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (yes)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (JavaDocs)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyMessage.java [hunk 1]
```diff
@@ -64,12 +65,53 @@
...
+	 * 		{@link NettyMessage} subclass ID
+	 * @param length
+	 * 		content length (or <tt>-1</tt> if unknown)
+	 *
+	 * @return a newly allocated direct buffer with header data written for {@link
+	 * NettyMessageDecoder}
+	 */
 	private static ByteBuf allocateBuffer(ByteBufAllocator allocator, byte id, int length) {
-		final ByteBuf buffer = length != 0 ? allocator.directBuffer(HEADER_LENGTH + length) : allocator.directBuffer();
+		Preconditions.checkArgument(length <= Integer.MAX_VALUE - HEADER_LENGTH);
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyMessage.java]: import static?

Comment (NicoK) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyMessage.java]: why not...

**Comment by pnowojski** [comment]
LGTM (besides "Conflicting files")

**Comment by NicoK** [comment]
of course, the last change added a conflict...rebased now

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyMessage.java [hunk 2]
```diff
@@ -67,12 +68,53 @@
 
 	// ------------------------------------------------------------------------
 
+	/**
+	 * Allocates a new (header and contents) buffer and adds some header information for the frame
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyMessage.java]: This method only allocates memory for the header and no contents (or contents is empty). Maybe we could update the JavaDoc accordingly.

**Pull request closed without merging**